plugins/experimental/server_push_preload/server_push_preload.cc

```diff
@@ -0,0 +1,141 @@
+
+#define PLUGIN_NAME "server_push_preload"
+#define PRELOAD_PARAM "rel=preload"
+#define NOPUSH_OPTION "nopush"
+
+using namespace std;
+using namespace atscppapi;
+
+static regex linkRegexp("<([^>]+)>;(.+)");
+
+namespace
+{
+GlobalPlugin *plugin;
+}
+
+class LinkServerPushPlugin : public GlobalPlugin
+{
+public:
+  LinkServerPushPlugin()
+  {
+    TSDebug(PLUGIN_NAME, "registering transaction hooks");
+    LinkServerPushPlugin::registerHook(HOOK_SEND_RESPONSE_HEADERS);
+  }
+
+  virtual void
+  handleSendResponseHeaders(Transaction &transaction)
+  {
+    serverPush(transaction);
+    transaction.resume();
+  }
+
+  void
+  serverPush(Transaction &transaction)
+  {
+    TSHttpTxn txnp = static_cast<TSHttpTxn>(transaction.getAtsHandle());
+    if (TSHttpTxnClientProtocolStackContains(txnp, "h2") == nullptr) {
+      return;
+    }
+
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
+      return;
+    }
+
+    Response &response = transaction.getClientResponse();
+    Headers &headers   = response.getHeaders();
+
+    const Url &clientUrl = request.getPristineUrl();
+
+    for (header_field_iterator it = headers.find("Link"); it != headers.end(); it.nextDup()) {
+      HeaderField field  = *it;
+      const string &link = field.str();
+
+      TSDebug(PLUGIN_NAME, "Parsing link header: %s", link.c_str());
+      smatch matches;
+
+      if (regex_search(link, matches, linkRegexp)) {
+        string url = matches[1].str();
+        TSDebug(PLUGIN_NAME, "Found link header match: %s", url.c_str());
+
+        unordered_map<string, bool> params = split(matches[2].str(), ';');
+        auto preload = params.find(PRELOAD_PARAM);
+        if (preload == params.end()) {
+          continue;
+        }
+
+        auto noPush = params.find(NOPUSH_OPTION);
+        if (noPush != params.end()) {
+          TSDebug(PLUGIN_NAME, "Skipping nopush link: %s", link.c_str());
+          continue;
+        }
+
+        Request request(url);
+        Url &linkUrl = request.getUrl();
+
+        if (linkUrl.getHost().empty()) {
+          linkUrl.setHost(clientUrl.getHost());
+          linkUrl.setScheme(clientUrl.getScheme());
+        }
+        string lu = linkUrl.getUrlString();
+        TSDebug(PLUGIN_NAME, "Push preloaded content: %s", lu.c_str());
+        TSHttpTxnServerPush(txnp, lu.c_str(), lu.length());
+      } else {
+        TSDebug(PLUGIN_NAME, "No match found for link header: %s", link.c_str());
+      }
+    }
+  }
+
+  unordered_map<string, bool>
+  split(const string &params, char delim)
+  {
+    stringstream ss(params);
+    string s;
+    unordered_map<string, bool> tokens;
+    while (getline(ss, s, delim)) {
+      s.erase(find_if(s.rbegin(), s.rend(), not1(std::ptr_fun<int, int>(isspace))).base(), s.end()); // trim left
+      s.erase(s.begin(), find_if(s.begin(), s.end(), not1(std::ptr_fun<int, int>(isspace))));        // trim right
+      tokens[s] = true;
```

Comment:
Set is enough, I updated the code. Sorry I'm too used to the lack of Set in Golang 😭